### PR TITLE
Commonized metadata infrastructure between native, js and native ide

### PR DESCRIPTION
**Pull request by alexander-gorshenev**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt [hunk 1]
```diff
@@ -160,13 +157,13 @@ private fun getFeaturedLibraries(
...
     featuredLibraryFiles: Set<File>,
-    resolvedLibraries: KonanLibraryResolveResult,
+    resolvedLibraries: KotlinLibraryResolveResult,
     reporter: FeaturedLibrariesReporter,
     allowDefaultLibs: Boolean
 ) : List<KonanLibrary> {
     val remainingFeaturedLibraries = featuredLibraryFiles.toMutableSet()
     val result = mutableListOf<KonanLibrary>()
-    val libraries = resolvedLibraries.getFullList(null)
+    val libraries = resolvedLibraries.getFullList(null) as List<KonanLibrary>
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt]: `KonanLibrary`?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/LegacyDescriptorUtils.kt [hunk 2]
```diff
@@ -161,23 +160,28 @@ internal val DeclarationDescriptor.isExpectMember: Boolean
 internal val DeclarationDescriptor.isSerializableExpectClass: Boolean
     get() = this is ClassDescriptor && ExpectedActualDeclarationChecker.shouldGenerateExpectClass(this)
 
-private fun sourceByIndex(descriptor: CallableMemberDescriptor, index: Int): SourceFile {
-    val fragment = descriptor.findPackage() as KonanPackageFragment
-    return fragment.sourceFileMap.sourceFile(index)
-}
+// TODO: temporary disabling. Need to figure out proper SourceFileMap and FileRegistry commonization.
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/LegacyDescriptorUtils.kt]: How it affects klibs?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/ContextUtils.kt [hunk 3]
```diff
@@ -7,21 +7,20 @@ package org.jetbrains.kotlin.backend.konan.llvm
...
 import org.jetbrains.kotlin.backend.konan.Context
 import org.jetbrains.kotlin.backend.konan.hash.GlobalHash
-import org.jetbrains.kotlin.backend.konan.ir.isReal
 import org.jetbrains.kotlin.backend.konan.ir.llvmSymbolOrigin
-import org.jetbrains.kotlin.descriptors.konan.CompiledKonanModuleOrigin
-import org.jetbrains.kotlin.descriptors.konan.CurrentKonanModuleOrigin
-import org.jetbrains.kotlin.descriptors.konan.DeserializedKonanModuleOrigin
+import org.jetbrains.kotlin.descriptors.konan.CompiledKlibModuleOrigin
+import org.jetbrains.kotlin.descriptors.konan.CurrentKlibModuleOrigin
+import org.jetbrains.kotlin.descriptors.konan.DeserializedKlibModuleOrigin
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/ContextUtils.kt]: Shall they actually be in `.konan` package at the end?

File: build.gradle [hunk 4]
```diff
@@ -65,6 +65,7 @@ ext {
     kotlinNativeUtilsModule="org.jetbrains.kotlin:kotlin-native-utils:${kotlinVersion}"
     kotlinUtilIoModule="org.jetbrains.kotlin:kotlin-util-io:${kotlinVersion}"
     kotlinUtilKlibModule="org.jetbrains.kotlin:kotlin-util-klib:${kotlinVersion}"
+    kotlinUtilKliMetadatabModule="org.jetbrains.kotlin:kotlin-util-klib-metadata:${kotlinVersion}"
```
Comment (olonho) [inline build.gradle]: Why can't we have everything klib related in single JAR?

File: cmd/run_konan [hunk 5]
```diff
@@ -94,8 +94,9 @@ UTILITIES_JAR="${KONAN_HOME}/konan/lib/utilities.jar"
 NATIVE_UTILS_JAR="${KONAN_HOME}/konan/lib/kotlin-native-utils.jar"
 UTIL_IO_JAR="${KONAN_HOME}/konan/lib/kotlin-util-io.jar"
 UTIL_KLIB_JAR="${KONAN_HOME}/konan/lib/kotlin-util-klib.jar"
+UTIL_KLIB_METADATA_JAR="${KONAN_HOME}/konan/lib/kotlin-util-klib-metadata.jar"
 TROVE_JAR="${KONAN_HOME}/konan/lib/trove4j.jar"
-KONAN_CLASSPATH="$KOTLIN_JAR:$KOTLIN_STDLIB_JAR:$KOTLIN_REFLECT_JAR:$KOTLIN_SCRIPT_RUNTIME_JAR:$INTEROP_JAR:$STUB_GENERATOR_JAR:$INTEROP_INDEXER_JAR:$KONAN_JAR:$KLIB_JAR:$UTILITIES_JAR:$SHARED_JAR:$EXTRACTED_METADATA_JAR:$EXTRACTED_SERIALIZER_JAR:$NATIVE_UTILS_JAR:$TROVE_JAR:$UTIL_IO_JAR:$UTIL_KLIB_JAR:$KOTLINX_CLI_JAR"
+KONAN_CLASSPATH="$KOTLIN_JAR:$KOTLIN_STDLIB_JAR:$KOTLIN_REFLECT_JAR:$KOTLIN_SCRIPT_RUNTIME_JAR:$INTEROP_JAR:$STUB_GENERATOR_JAR:$INTEROP_INDEXER_JAR:$KONAN_JAR:$KLIB_JAR:$UTILITIES_JAR:$SHARED_JAR:$EXTRACTED_METADATA_JAR:$EXTRACTED_SERIALIZER_JAR:$NATIVE_UTILS_JAR:$TROVE_JAR:$UTIL_IO_JAR:$UTIL_KLIB_JAR:$UTIL_KLIB_METADATA_JAR:$KOTLINX_CLI_JAR"
```
Comment (olonho) [inline cmd/run_konan]: Same, list here is ridiculously long and shall be shortened.

File: gradle.properties [hunk 6]
```diff
@@ -29,4 +29,5 @@ org.gradle.jvmargs='-Dfile.encoding=UTF-8'
 org.gradle.workers.max=4
 
 # Uncomment to enable composite build
-#kotlinProjectPath=<insert the path to Kotlin project root here>
\ No newline at end of file
+#kotlinProjectPath=<insert the path to Kotlin project root here>
+kotlinProjectPath=/Users/jetbrains/kotlin-native/kotlin
```
Comment (olonho) [inline gradle.properties]: Not for merge.

File: shared/src/library/kotlin/org/jetbrains/kotlin/konan/library/SearchPathResolver.kt [hunk 7]
```diff
@@ -6,48 +6,32 @@ import org.jetbrains.kotlin.konan.library.impl.KonanLibraryImpl
...
-        compatibleCompilerVersions: List<KonanVersion> = emptyList(),
-        logger: Logger = DummyLogger,
-        skipCurrentDir: Boolean = false
-): SearchPathResolverWithTarget = KonanLibraryProperResolver(
+    repositories: List<String>,
+    directLibs: List<String>,
+    target: KonanTarget,
+    distribution: Distribution,
+    compatibleCompilerVersions: List<KonanVersion> = emptyList(),
+    logger: Logger = DummyLogger,
```
Comment (olonho) [inline shared/src/library/kotlin/org/jetbrains/kotlin/konan/library/SearchPathResolver.kt]: BTW, shall it be indeed default logger?

File: build.gradle (hunk 4 above)
Comment (ddolovov) [reply build.gradle]: +1, likely nobody needs KLIB reader w/o metadata

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt (hunk 1 above)
Comment (alexander-gorshenev) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt]: Это место, где для нас становится важным, что библиотека может иметь konan specific extensions. Но пока никакого кода, который умеет различать библиотеки в этом смысле у нас нет.

File: build.gradle (hunk 4 above)
Comment (alexander-gorshenev) [reply build.gradle]: util-klib-metadata has all the compiler internals such as descriptors as its dependency. 
util-klib has only ByteArray and String in its interfaces. 
A CLI tool to pack sources to the library would only need util-klib, but not util-klib-metadata.

File: gradle.properties (hunk 6 above)
Comment (alexander-gorshenev) [reply gradle.properties]: fixed

File: shared/src/library/kotlin/org/jetbrains/kotlin/konan/library/SearchPathResolver.kt (hunk 7 above)
Comment (alexander-gorshenev) [reply shared/src/library/kotlin/org/jetbrains/kotlin/konan/library/SearchPathResolver.kt]: This is just a default argument value. It allows klib infrastructure to be used in the absence of  Kotlin compiler infrastructure (the message collector, etc). For the compiler we use an object based on the message collector

**Pull request merged**